pkg/registry/core/pod/strategy.go

```diff
@@ -569,3 +572,120 @@ func validateContainer(container string, pod *api.Pod) (string, error) {
 
 	return container, nil
 }
+
+// applySeccompVersionSkew implements the version skew behavior described in:
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/20190717-seccomp-ga.md#version-skew-strategy
+func applySeccompVersionSkew(pod *api.Pod) {
+	// initialize parents if necessary
+	if pod.Annotations == nil {
+		pod.Annotations = map[string]string{}
+	}
+
+	annotation, hasAnnotation := pod.Annotations[v1.SeccompPodAnnotationKey]
+	field, hasField := (*api.SeccompProfile)(nil), false
+	if pod.Spec.SecurityContext != nil && pod.Spec.SecurityContext.SeccompProfile != nil {
+		field = pod.Spec.SecurityContext.SeccompProfile
+		hasField = true
+	}
+
+	if hasAnnotation != hasField {
+		// assign computed values
+		newAnnotation, newField := syncSeccompAnnotationAndField(annotation, field)
+		if newAnnotation != "" {
+			pod.Annotations[v1.SeccompPodAnnotationKey] = newAnnotation
+		}
+		if newField != nil {
+			if pod.Spec.SecurityContext == nil {
+				pod.Spec.SecurityContext = &api.PodSecurityContext{}
+			}
+			pod.Spec.SecurityContext.SeccompProfile = newField
+		}
+	}
+
+	// Handle the containers of the pod
+	podutil.VisitContainers(&pod.Spec, podutil.AllFeatureEnabledContainers(),
+		func(ctr *api.Container, _ podutil.ContainerType) bool {
+			// find the annotation value
+			var (
+				annotation    string
+				hasAnnotation bool
+			)
+			for key, value := range pod.Annotations {
+				if key == api.SeccompContainerAnnotationKeyPrefix+ctr.Name {
+					annotation = value
+					hasAnnotation = true
+					break
+				}
+			}
+
+			// find the field
+			field, hasField := (*api.SeccompProfile)(nil), false
+			if ctr.SecurityContext != nil && ctr.SecurityContext.SeccompProfile != nil {
+				field = ctr.SecurityContext.SeccompProfile
+				hasField = true
+			}
+
+			if hasAnnotation != hasField {
+				// initialize parents if necessary
+				// assign computed values
+				newAnnotation, newField := syncSeccompAnnotationAndField(annotation, field)
+				if newAnnotation != "" {
+					pod.Annotations[v1.SeccompContainerAnnotationKeyPrefix+ctr.Name] = newAnnotation
+				}
+				if newField != nil {
+					if ctr.SecurityContext == nil {
+						ctr.SecurityContext = &api.SecurityContext{}
+					}
+					ctr.SecurityContext.SeccompProfile = newField
+
+				}
+			}
+
+			return true
+		})
+}
+
+// syncSeccompAnnotationAndField takes a pod annotation and a seccomp profile
+// field and returns the synchronized result: The annotation will be converted
+// in an appropriate profile and vice versa.
+func syncSeccompAnnotationAndField(annotation string, field *api.SeccompProfile) (string, *api.SeccompProfile) {
+	// If only seccomp fields are specified, add the corresponding annotations.
+	// This ensures that the fields are enforced even if the node version
+	// trails the API version
+	if field != nil && annotation == "" {
+		switch field.Type {
+		case api.SeccompProfileTypeUnconfined:
+			return v1.SeccompProfileNameUnconfined, nil
```

Comment:
I had envisioned returning both the annotation value and the field value in all cases. Doing it this way is fine, but if this is going to be implemented in a way that only returns one of the values, it would probably make more sense as two separate methods:

```go
func seccompAnnotationForField(field *api.SeccompProfile) string {
... assume field is non-nil ...
}

func seccompFieldForAnnotation(annotation string) *api.SeccompProfile {
... assume annotation is non-empty ...
}
```

and be called like this:
```go
switch {
  case hasAnnotation && !hasField:
    if pod.Spec.SecurityContext == nil {
      pod.Spec.SecurityContext = &api.PodSecurityContext{}
    }
    pod.Spec.SecurityContext.SeccompProfile = seccompFieldForAnnotation(annotation)

  case hasField && !hasAnnotation:
    if pod.Annotations == nil {
      pod.Annotations = map[string]string{}
    }
    pod.Annotations[key] = seccompAnnotationForField(field)
}
```

Reply:
I split-up the function into two. They can still return `nil` or `""` if the localhost profile is invalid, so I have to check for the result again.